### DISCUSSION TITLE
JMustache 1.16 released and now supports Parents and Blocks

### DISCUSSION THREAD

**Discussion by agentgt** [mustache/spec#182] [Show and tell]
While I'm the primary author of JStachio I have been helping [JMustache](https://github.com/samskivert/jmustache) modernize.

It now supports Parents and Blocks and passes all the of the 1.3 spec tests (including correct whitespace).

@jgonggrijp Wontache playground was super helpful for me w/ whitespace checking. That was by far the hardest part. After doing this twice now I'm not sure why it is but it is totally worth it to get it correct.  While I could have used JStachio for confirmation it is a fair amount of ceremony compared to the ease that is Wontache Playground.

Speaking of Wontache: JMustache is very similar on how readable and small the codebase is .... (cough especially compared to the beast of compiler that is JStachio).

I plan on adding experimental power filters #153 in a separate branch as it is much easier to do that in JMustache at the moment.

**Comment by jgonggrijp** [comment]
Glad the Wontache playground has been useful to you!

Does JMustache also implement #131?

I can think of a few reasons why whitespace handling is tricky:

- The rules in the spec are rather implicit. You have to largely infer them from the tests and it is easy to make wrong guesses along the way.
- There is a strong bias towards doing what a template author would intuitively expect. People's intuitions often require intricate logic to reproduce faithfully.
- However, the spec makes exceptions to human intuition when there is an alternative with potentially superior performance. Hence, intuition can still be misleading. I personally found this test the most surprising, though I fully agree with it now that I understand it:

https://github.com/mustache/spec/blob/83b221aec50adb91eeaa935ba7bd51cd792daece/specs/partials.yml#L82-L100

After implementing the rules correctly, they seemed less complicated to me than before. I think this summarizes the rules:

- A tag that clears on both sides and that is not an interpolation tag, is standalone. #131 generalizes this to parent tag pairs and block tag pairs.
- If a tag is standalone, its entire line should be considered part of the tag. In other words, leading and trailing whitespace around the tag is *not* part of the literal template content. This includes the final newline character if there is one. Again, #131 generalizes this to parent tag pairs and block tag pairs.
- A block behaves like a template. If its contents are indented, this indentation is not really part of the template. #131 makes this explicit.
- Indentation may be added to a template *before* rendering.
